### ISSUE TITLE
How to get ownership of data back from ImageSurface

### ISSUE THREAD

**Issue by jimpo** [gtk-rs/cairo#335]
## Summary

You can quickly create an `ImageSurface` from some byte array using `ImageSurface::create_for_data`. However, I don't see a way to get ownership of that byte array back from the `ImageSurface` without performing a copy. I would like to figure out a way to avoid copying my image data.

## More context

I am working on an application that uses a background thread to render an image and uses double buffering to quickly update the main display. My approach is inspired by [this tutorial](https://www.cairographics.org/threaded_animation_with_cairo/), but updated for compatibility with GTK+ 3 and obviously using the Rust APIs (and ideally just safe Rust). I am using a `Vec<u8>` as a pixel map which I can pass between threads, and from it construct an `ImageSurface` using `ImageSurface::create_for_data`. The problem is, after drawing on the `ImageSurface`, I'd like to be able to take back ownership of my pixel map without cloning the data into a new `Vec` from the `ImageSurfaceData` returned by `ImageSurface::get_data`. Right now, it seems my options are to either to clone the data or drop into `unsafe`.

## Possible approaches

I see from [the code](https://github.com/gtk-rs/cairo/blob/master/src/image_surface.rs#L62) that the byte array is stored as user data on the created surface. However, this user data key is not accessible publicly. If it were, one could presumably get the `Rc` handle from the user data, drop the `ImageSurface` and use `Rc::try_unwrap` and `Box::downcast` to recover the byte array. That's gross though. An idea that's only slightly better at best would be to return the `Rc<Box<dyn AsMut<u8>>>` from `create_for_data` along with the `ImageSurface`, which at least doesn't expose the user data key.

Probably the cleanest API I can think of would be a new generic struct like `ImageSurfaceWithData<D: AsMut<u8>>` which is `Deref<Target = ImageSurface>`. But that starts to deviate even more from the C API. That's basically the API that I will likely implement myself with `unsafe` if there is no upstream solution.

Any thoughts on if there already exists an approach that I missed or whether this is worth addressing?

**Comment by sdroege** [comment]
Take a look at https://github.com/gtk-rs/examples/blob/master/src/bin/cairo_threads.rs for one possible safe solution. Otherwise you'll have to go with unsafe code like in https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/f40821ba662e6ba29ddb72915ea30ee4191da3ea/examples/src/bin/overlay-composition.rs#L176 .

The problem with your approaches is that the `ImageSurface` is reference counted, so even if your reference of the surface is dropped that doesn't yet mean that the surface is really gone and nothing is making use of the data anymore.

**Comment by jimpo** [comment]
Thanks for the references!

But couldn't a method like `finish_output_data` check the number of references on the surface the same way `get_data` currently does and error if it is not an exclusive reference?

**Comment by sdroege** [comment]
That would indeed be an option, yes. Do you want to provide a PR for such a thing?

**Comment by jimpo** [comment]
Yep, I'll put something up for comment within the next few days.

**Comment by sdroege** [comment]
@jimpo Any updates here?

**Comment by jimpo** [comment]
No, sorry, I haven't been working on it. I just worked around it with unsafe. Feel free to close the issue and I may or may not submit a PR at some point.

**Comment by sdroege** [comment]
No need to close it, but like this it's documented that nobody is working on it right now so maybe someone else takes it up :) Thanks!
